### PR TITLE
Fix/remove bams storage

### PR DESCRIPTION
**Pull request by KateSakharova**
Those fixes would include remove of some intermediate heavy files
- removed diamond alignments from CAT step
- removed bams from decontamination step
- added `samtools bam2fq` into alignment 

TODO: we still have 1 place where there are bam files: alignment before METABAT2_JGISUMMARIZEBAMCONTIGDEPTHS. 
Possible solutions can be:
- add metabat to bwa container and combine depth calcutation to alignment
- calculate depth with `samtools` but results would be different from jgi script 

UPD: CONCOCT also requires bam files
https://concoct.readthedocs.io/en/latest/scripts/concoct_coverage_table.html

### CODE REVIEW

File: config/codon.config [hunk 1]
```diff
@@ -14,7 +14,12 @@ workDir = params.workdir
 
 process {
 
-    withName: ALIGNMENT {
+    withName: ALIGNMENT_BAM {
+        cpus = 16
+        memory     = { 50.GB * task.attempt }
+        maxRetries    = 5
```
Comment (mberacochea) [inline config/codon.config:20]: Loads of retries, we should capture the stats from successful runs to tweak this a bit. As that many retries will bloat the nf workdir

File: modules/local/align_bwa/main.nf [hunk 2]
```diff
@@ -68,6 +68,84 @@ process ALIGNMENT {
...
+        bwa-mem2: \$(bwa-mem2 version 2> /dev/null)
+        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
+    END_VERSIONS
+    """
+}
+
+process ALIGNMENT_READS {
+
+    // module does alignment with chosen arguments and generates FASTQ.GZ files as output
+    // that module was done to remove storage of heavy bam files
```
Comment (mberacochea) [inline modules/local/align_bwa/main.nf:82]: A suggestion
```suggestion
    /*
    This module aligns reads to the reference using specified arguments and produces output in the form of FASTQ.GZ files. 
    Once the alignment is completed, the BAM files are removed to save disk space.
    */
```

File: modules/local/align_bwa/main.nf [hunk 3]
```diff
@@ -25,8 +25,8 @@ process INDEX_FASTA {
     """
 }
 
-process ALIGNMENT {
-
+process ALIGNMENT_BAM {
+    // module does alignment with chosen arguments and generates BAM and BAM.BAI files as output
```
Comment (mberacochea) [inline modules/local/align_bwa/main.nf:29]:
```suggestion
    /*
    This module aligns reads to the reference using specified arguments and produces output in the form of FASTQ.GZ, BAM and BAM.BAI files.
    */
```

File: modules/local/cat/cat/main.nf [hunk 4]
```diff
@@ -34,6 +34,13 @@ process CAT {
     CAT summarise -c ${bin_fa} \
     -i ${meta.id}.contig2classification.official_names.txt -o ${meta.id}.summary.txt
 
+    echo "Remove diamond alignment"
+    rm "${meta.id}.alignment.diamond"
+
+    echo "Remove predicted proteins"
+    rm "${meta.id}.predicted_proteins.faa" || true
+    rm "${meta.id}.predicted_proteins.gff" || true
```
Comment (mberacochea) [inline modules/local/cat/cat/main.nf:42]:
```suggestion
    rm -f "${meta.id}.predicted_proteins.faa"
    rm -f "${meta.id}.predicted_proteins.gff"
```

File: modules/local/cat/cat/main.nf [hunk 5]
```diff
@@ -34,6 +34,13 @@ process CAT {
     CAT summarise -c ${bin_fa} \
     -i ${meta.id}.contig2classification.official_names.txt -o ${meta.id}.summary.txt
 
+    echo "Remove diamond alignment"
+    rm "${meta.id}.alignment.diamond"
```
Comment (mberacochea) [inline modules/local/cat/cat/main.nf:38]:
```suggestion
    rm -f "${meta.id}.alignment.diamond"
```
just in case

**Review by mberacochea — APPROVED** [review]
Good stuff, thanks Kate. I think this will save loads of storage.

I left you some minor comments and suggestions but overall it looks good to me :)

**Comment by mberacochea** [comment]
> Those fixes would include remove of some intermediate heavy files
> 
>     * removed diamond alignments from CAT step
> 
>     * removed bams from decontamination step
> 
>     * added `samtools bam2fq` into alignment
> 
> 
> TODO: we still have 1 place where there are bam files: alignment before METABAT2_JGISUMMARIZEBAMCONTIGDEPTHS. Possible solutions can be:
> 
>     * add metabat to bwa container and combine depth calcutation to alignment
> 
>     * calculate depth with `samtools` but results would be different from jgi script

I think adding metabat to the bwa container to combine the depth calculation is the best way, moving from the jgi to the samtools one would require some exploration and benchmarking

**Review by Ales-ibt — APPROVED** [review]
I like the idea of having sam2fastq in the same module where we run bwamem2. Also, removing the bam files for decontamination will definitely help with the storage. Maybe we can use this idea in the nf-core decontamination module as well adding an idxstats (samtools) step

**Pull request merged**